### PR TITLE
Authentication Support for Java & Go SDKs

### LINKED ISSUE #950: Authentication & Authorization Support for Java and Go SDKs
### Background
Feast has authentication & authorization support completed for the following components:
-  Core #793 
-  Serving #865 
- Python SDK (#763 & #865)

### Problem
Authentication & Authorization support does not extend to Java and Golang SDK:
- This blocks administrators from having a full authentication and authorization rollout if they have users using Java and Golang SDKs.

### What needs to be done
Extend Authentication & Authorization support to
- Java SDK
- Golang SDK

such that they can access Feast Core and Serving with Authentication & Authorization enabled.

### PR DESCRIPTION
**Pull request by mrzzy**
**What this PR does / why we need it**:

Implements support for authentication to Go and Java SDKs, which is required for interacting with a authentication enabled Feast Serving.
- Update Feast clients to attach OIDC token used for authentication into metadata when enabled.
- Allows users to provide credentials via Google, OAuth Client Credentials request, or a Static token.
  - Google: Uses user provided Service Account Credentials (ie JSON) to obtain token.
  - OAuth Client Credentials: Makes a client credentials request to obtain token.
  - Static: Uses a user provided static token.

**Which issue(s) this PR fixes**:

Fixes #950 

**Does this PR introduce a user-facing change?**:

```release-note
Adds authentication support to Go and Java SDKs via Google, OAuth Client Credentials request, or a Static token.
```

### CODE REVIEW

File: .prow/config.yaml [hunk 1]
```diff
@@ -146,7 +146,7 @@ presubmits:
     always_run: true
     spec:
       containers:
-      - image: maven:3.6-jdk-11
+      - image: gcr.io/kf-feast/feast-ci:latest
```
Comment (woop) [inline .prow/config.yaml:149]: Can you explain these changes?

File: common/src/main/java/feast/common/auth/credentials/GoogleAuthCredentials.java [hunk 2]
```diff
@@ -28,18 +28,25 @@
...
+ * passed as authorization bearer token when making calls.
  */
 public class GoogleAuthCredentials extends CallCredentials {
   private final IdTokenCredentials credentials;
   private static final String BEARER_TYPE = "Bearer";
   private static final Metadata.Key<String> AUTHORIZATION_METADATA_KEY =
       Metadata.Key.of("Authorization", ASCII_STRING_MARSHALLER);
 
+  /**
+   * Constructa new GoogleAuthCredentials with given options.
```
Comment (woop) [inline common/src/main/java/feast/common/auth/credentials/GoogleAuthCredentials.java:44]: Typo

File: infra/scripts/setup-common-functions.sh [hunk 3]
```diff
@@ -10,14 +10,17 @@ install_test_tools() {
...
+  print_banner "Installing Google Cloud SDK. (GOOGLE_APPLICATION_CREDENTIALS: '$GOOGLE_APPLICATION_CREDENTIALS')_"
   if [[ ! $(command -v gsutil) ]]; then
     CURRENT_DIR=$(dirname "$BASH_SOURCE")
-    . "${CURRENT_DIR}"/install-google-cloud-sdk.sh
+    # Activate service account credentials if specified.
+    if [ -z $GOOGLE_APPLICATION_CREDENTIALS ]
+    then
+      . "${CURRENT_DIR}"/install-google-cloud-sdk.sh
+    else
+      . "${CURRENT_DIR}"/install-google-cloud-sdk.sh --with-key-file $GOOGLE_APPLICATION_CREDENTIALS
```
Comment (woop) [inline infra/scripts/setup-common-functions.sh:21]: Is there a reason to keep that installation script separate and not move it into this functions script? This just seems like a lot of brittle scripts

File: infra/scripts/test-end-to-end.sh [hunk 4]
```diff
@@ -5,11 +5,13 @@ set -o pipefail
 [[ $1 == "True" ]] && ENABLE_AUTH="true" || ENABLE_AUTH="false"
 echo "Authenication enabled : ${ENABLE_AUTH}"
 
-test -z ${GOOGLE_APPLICATION_CREDENTIALS} && GOOGLE_APPLICATION_CREDENTIALS="/etc/gcloud/service-account.json"
+test -z ${GOOGLE_APPLICATION_CREDENTIALS} && GOOGLE_APPLICATION_CREDENTIALS=""
```
Comment (woop) [inline infra/scripts/test-end-to-end.sh:8]: Why are we clearing that configuration? Isn't that a default?

File: infra/scripts/test-end-to-end.sh [hunk 5]
```diff
@@ -5,11 +5,13 @@ set -o pipefail
...
 test -z ${SKIP_BUILD_JARS} && SKIP_BUILD_JARS="false"
 test -z ${GOOGLE_CLOUD_PROJECT} && GOOGLE_CLOUD_PROJECT="kf-feast"
 test -z ${TEMP_BUCKET} && TEMP_BUCKET="feast-templocation-kf-feast"
 test -z ${JOBS_STAGING_LOCATION} && JOBS_STAGING_LOCATION="gs://${TEMP_BUCKET}/staging-location"
+# Target SDK to run E2E tests with. Supported values: 'python', 'java', 'go'
```
Comment (woop) [inline infra/scripts/test-end-to-end.sh:13]: I dont think there is a plan to have full support for e2e tests or all functionality with other SDKs. Go and Java will probably only have limited functionality compared to the Python library, especially if we move towards PySpark.

Also, even if we had full compatibility between the SDKs of all functionality, we would still want to use a single e2e test framework like Pytest ideally, just to keep things consistent.

File: tests/e2e/java/pom.xml [hunk 6]
```diff
@@ -0,0 +1,105 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (woop) [inline tests/e2e/java/pom.xml:1]: It's not clear why we need a new test module. Why cant the tests be integration tests without the existing SDK module?

File: tests/e2e/go/e2e_test.go [hunk 7]
```diff
@@ -0,0 +1,119 @@
+package main
```
Comment (woop) [inline tests/e2e/go/e2e_test.go:1]: Again, its not clear why we need more than just Pytest for e2e. We can use https://github.com/testcontainers/testcontainers-go if we need integration tests from the existing test suites.

File: .prow/config.yaml (hunk 1 above)
Comment (mrzzy) [reply .prow/config.yaml:149]: This provides the test with `go` executable that it needs to run `go test`.

File: infra/scripts/test-end-to-end.sh (hunk 4 above)
Comment (mrzzy) [reply infra/scripts/test-end-to-end.sh:8]: `install_gcloud_sdk()` tries to activate service account credential if this variable is set. Under the current `.prow/config.yaml` `test-end-to-end` does not expose the service account key.json.

**Comment by mrzzy** [comment]
Reverted the E2E test changes.

File: common/src/main/java/feast/common/auth/credentials/GoogleAuthCredentials.java (hunk 2 above)
Comment (mrzzy) [reply common/src/main/java/feast/common/auth/credentials/GoogleAuthCredentials.java:44]: fixed.

File: infra/scripts/test-end-to-end.sh (hunk 5 above)
Comment (mrzzy) [reply infra/scripts/test-end-to-end.sh]: Reverted.

File: infra/scripts/test-end-to-end.sh (hunk 4 above)
Comment (mrzzy) [reply infra/scripts/test-end-to-end.sh]: Reverted.

File: common/src/main/java/feast/common/auth/credentials/GoogleAuthCredentials.java (hunk 2 above)
Comment (mrzzy) [reply common/src/main/java/feast/common/auth/credentials/GoogleAuthCredentials.java]: Reverted.

File: .prow/config.yaml (hunk 1 above)
Comment (mrzzy) [reply .prow/config.yaml]: Reverted.

File: tests/e2e/java/pom.xml (hunk 6 above)
Comment (mrzzy) [reply tests/e2e/java/pom.xml]: Reverted.

File: tests/e2e/go/e2e_test.go (hunk 7 above)
Comment (mrzzy) [reply tests/e2e/go/e2e_test.go]: Reverted.

**Review by pyalex — APPROVED** [review]

**Pull request merged**